File: 1514-path-with-maximum-probability/1514-path-with-maximum-probability.cpp
class Solution {
public:
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& succProb, int start, int end) {
        vector<vector<pair<int, double>>> adjList(n);
        int i = 0;
        for(auto e: edges){
            int u = e[0], v = e[1];
            adjList[u].push_back({v, succProb[i]});
            adjList[v].push_back({u, succProb[i++]});
        }
        
        vector<double> prob(n, 0.0);    // cost to reach a node (here: probability)
        prob[start] = 1.0;
        queue <int> q;
        q.push(start);
        while (!q.empty()){
            int cur = q.front();
            q.pop();
            for(auto nbr: adjList[cur]){
                double prob_new = prob[cur] * nbr.second;
                if(prob_new > prob[nbr.first]){
                    prob[nbr.first] = prob_new;
                    // cout << cur << "---" << nbr.first << "<-" << prob_new << "\n";
                    q.push(nbr.first);
                }
            }
        }
        return prob[end];
    }
};

File: 1514-path-with-maximum-probability/1514-path-with-maximum-probability.java
class Solution {
    private record Node(int idx, double prob) {
    }

    public double maxProbability(int n, int[][] edges, double[] succProb, int start_node, int end_node) {
        List<Node>[] adj = new List[n];
        boolean[] vis = new boolean[n];
        for (int i = 0; i < n; i++)
            adj[i] = new ArrayList<>();
        for (int i = 0; i < edges.length; i++) {
            adj[edges[i][0]].add(new Node(edges[i][1], succProb[i]));
            adj[edges[i][1]].add(new Node(edges[i][0], succProb[i]));
        }
        return dijkstra(adj, vis, start_node, end_node);
    }

    private double dijkstra(List<Node>[] adj, boolean[] vis, int src, int dest) {
        Queue<Node> pq = new PriorityQueue<>((a, b) -> Double.compare(b.prob, a.prob));
        pq.offer(new Node(src, 1d));
        while (!pq.isEmpty()) {
            Node best = pq.poll();
            // System.out.println(best);
            vis[best.idx] = true;
            if (best.idx == dest)
                return best.prob;
            for (Node nbr : adj[best.idx])
                if(!vis[nbr.idx])
                    pq.offer(new Node(nbr.idx, best.prob * nbr.prob));
        }
        return 0d;
    }
}